### PR TITLE
feat: add `Par` expression pt. 1

### PR DESCRIPTION
**Pull request by jaschdoc**
Related to #4286

### CODE REVIEW

**Comment by magnus-madsen** [comment]
I have a proposal: Maybe it is easier to parse everything as just `Par exp`. Then through all the phases we just have `case class Par(exp)` until `Lowring` (?) where we do the rewrite.

**Comment by jaschdoc** [comment]
RFR

File: main/src/ca/uwaterloo/flix/language/phase/Typer.scala [hunk 1]
```diff
@@ -1595,6 +1595,8 @@ object Typer {
           resultEff = eff
         } yield (constrs, resultTyp, resultPur, resultEff)
 
+      case KindedAst.Expression.Par(exp, tvar, pvar, evar, loc) => ???
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Typer.scala:1598]: Here we should just call recursively on exp and return its type, purity, effect.

File: main/src/ca/uwaterloo/flix/language/phase/Typer.scala [hunk 2]
```diff
@@ -2216,6 +2218,8 @@ object Typer {
         val eff = e.eff
         TypedAst.Expression.Spawn(e, tpe, pur, eff, loc)
 
+      case KindedAst.Expression.Par(exp, tvar, pvar, evar, loc) => ???
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Typer.scala:2221]: We will need a TypedAst.Exp.Par.

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala [hunk 3]
```diff
@@ -666,7 +668,8 @@ object Resolver {
             case NamedAst.DefOrSig.Sig(sig) => visitApplySig(app, sig, exps, region, innerLoc, outerLoc)
           }
 
-        case app@NamedAst.Expression.Apply(_, _, _) => visitApply(app, region)
+        case app@NamedAst.Expression.Apply(_, _, _) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:671]: Remove app@

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala [hunk 4]
```diff
@@ -1504,6 +1504,12 @@ object Weeder {
         case e => WeededAst.Expression.Spawn(e, mkSL(sp1, sp2))
       }
 
+    case ParsedAst.Expression.Par(sp1, exp, sp2) =>
+      mapN(visitExp(exp, senv)) {
+        e =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:1509]: One line

File: main/src/ca/uwaterloo/flix/language/phase/Kinder.scala [hunk 5]
```diff
@@ -715,6 +715,15 @@ object Kinder {
         exp => KindedAst.Expression.Spawn(exp, loc)
       }
 
+    case ResolvedAst.Expression.Par(exp, loc) =>
+      mapN(visitExp(exp, kenv0, senv, taenv, henv0, root)) {
+        e =>
+          val tvar = Type.freshVar(Kind.Star, loc.asSynthetic)
+          val pvar = Type.freshVar(Kind.Bool, loc.asSynthetic)
+          val evar = Type.freshVar(Kind.Effect, loc.asSynthetic)
+          KindedAst.Expression.Par(e, tvar, pvar, evar, loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Kinder.scala:724]: We will not need any fresh variables. Just the expression :)

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala (hunk 3 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:671]: This was already here. It's needed for the function call to `visitApply` :)

**Comment by jaschdoc** [comment]
Should we require that `exp` must be pure?

**Comment by magnus-madsen** [comment]
> Should we require that `exp` must be pure?

Not for now. We allow anything.

**Comment by jaschdoc** [comment]
RFR

File: main/src/ca/uwaterloo/flix/language/phase/Monomorph.scala [hunk 6]
```diff
@@ -619,6 +619,9 @@ object Monomorph {
       case Expression.FixpointProject(_, _, _, _, _, loc) =>
         throw InternalCompilerException(s"Unexpected expression near: ${loc.format}.")
 
+      case Expression.Par(_, loc) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Monomorph.scala:622]: This should be moved after Spawn or alternatively to the bottom.

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala (hunk 3 above)
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:671]: Remove app@ binding

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:671]: Wait, sorry, I am confused. What is happening here.

File: main/src/ca/uwaterloo/flix/language/phase/Monomorph.scala (hunk 6 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Monomorph.scala]: See https://github.com/flix/flix/pull/4327#discussion_r928801498

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala (hunk 3 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:671]: I guess the code was already like that.

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:671]: https://github.com/flix/flix/pull/4327#discussion_r928801498

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:671]: Yeah exactly :+1:

**Comment by jaschdoc** [comment]
Love how it's unable to build a jar :joy:

**Comment by jaschdoc** [comment]
RFR

**Comment by magnus-madsen** [comment]
Great! Now I think all the work that remains is in the Lowering.

**Pull request merged**